### PR TITLE
Issue #439 Strip out double quotes in ARG value

### PR DESCRIPTION
**Pull request by cvgw**
Fixes [#439](https://github.com/GoogleContainerTools/kaniko/issues/439)

**Description**

Strip out double quotes enclosing ARG value after parsing Dockerfile. This brings Kaniko into parity with Docker for handling ARG values. This prevents failures where things like image names would not parse correctly due to the double quotes remaining.

**Submitter Checklist**

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [unit tests](../DEVELOPMENT.md#creating-a-pr)
- [x] Adds integration tests if needed.

_See [the contribution guide](../CONTRIBUTING.md) for more details._

**Reviewer Notes**

- [ ] The code flow looks good. 
- [ ] Unit tests and or integration tests added.

**Release Notes**

Describe any changes here so maintainer can include it in the release notes, or delete this block.

```
- Strip double quotes enclosing the value of ARG commands before substituting those values in during shell expansion
```

### CODE REVIEW

File: pkg/dockerfile/dockerfile.go [hunk 1]
```diff
@@ -111,9 +111,33 @@ func Parse(b []byte) ([]instructions.Stage, []instructions.ArgCommand, error) {
...
 		return nil, nil, err
 	}
+
+	metaArgs = stripEnclosingDoubleQuotes(metaArgs)
+
 	return stages, metaArgs, nil
 }
 
+// stripEnclosingDoubleQuotes removes double quotes enclosing the value of each instructions.ArgCommand in a slice
+func stripEnclosingDoubleQuotes(metaArgs []instructions.ArgCommand) []instructions.ArgCommand {
```
Comment (tejal29) [inline pkg/dockerfile/dockerfile.go]: We can add table driven unit tests here
1. value like`"foo"`
2. For empty value `""` ? What happens then?
   maybe we shd keep quotes as is then?
3.  value `foo`
4.  invalid value `"foo` shd probably error out?

Comment (tejal29) [reply pkg/dockerfile/dockerfile.go]: Some other test cases to add would be
5. When `metaArgs` is empty
6. Multiple `mataArgs` instructions have `""`. (All get resolved)

Comment (priyawadhwa) [reply pkg/dockerfile/dockerfile.go]: just something to consider, something like `\"foo\"` would probably need to maintain the quotes

**Comment by cvgw** [comment]
> I didn't look thoroughly at this, but this may be what docker uses to [resolve quotes](https://github.com/moby/buildkit/blob/master/frontend/dockerfile/shell/lex.go#L13-L19), which we could potentially use as well to ensure we respond to this as docker does

@priyawadhwa kaniko is currently [using that same package](https://github.com/GoogleContainerTools/kaniko/blob/master/pkg/util/command_util.go#L60) to do the variable expansion. I was surprised that it doesn't seem to be handling the quotes, but I didn't see anything in the documentation referring to that. I did notice that in a recent version (later than what kaniko seems to be using) that `RawQuotes` field was added. Perhaps that is related?

File: pkg/dockerfile/dockerfile.go (hunk 1 above)
Comment (cvgw) [reply pkg/dockerfile/dockerfile.go]: > We can add table driven unit tests here for this function only
> 
> 1. value like`"foo"`
> 2. For empty value `""` ? What happens then?
>    maybe we shd keep quotes as is then?
> 3. value `foo`
> 4. invalid value `"foo` shd probably error out?

2. Docker seems to strip out the quotes and treat it as a valid, but blank string
4. Yup, Docker also seems to error when either quote isn't matched

We also might want to handle single quotes as docker seems to handle those as well

Comment (cvgw) [reply pkg/dockerfile/dockerfile.go]: > just something to consider, something like `\"foo\"` would probably need to maintain the quotes

Ya, it seems that Docker does not remove the quotes in that case which will result in an error for things like image name

**Comment by priyawadhwa** [comment]
>  I did notice that in a recent version (later than what kaniko seems to be using) that RawQuotes field was added. Perhaps that is related?

Yah that sounds like it could be! We could definitely bump up the version if you think that would work

**Comment by cvgw** [comment]
> > I did notice that in a recent version (later than what kaniko seems to be using) that RawQuotes field was added. Perhaps that is related?
> 
> Yah that sounds like it could be! We could definitely bump up the version if you think that would work

Unfortunately doesn't seem to be related
```
package buildkit_quotes_test
  
  import (
          "testing"
  
          "github.com/moby/buildkit/frontend/dockerfile/parser"
          "github.com/moby/buildkit/frontend/dockerfile/shell"
  )
  
  func TestWithDoubleQuotes(t *testing.T) {
          envs := []string{"MEOW=purr", "VAL=\"World\""}
  
          lex := shell.NewLex(parser.DefaultEscapeToken)
          val, err := lex.ProcessWord("Hello ${VAL}", envs)
          if err != nil {
                  t.Error(err)
          }   
  
          if val != "Hello World" {
                  t.Errorf("Expected Hello World but got %s", val)
          }   
  }
  
  func TestWithDoubleQuotesRawQuotesTrue(t *testing.T) {
          envs := []string{"MEOW=purr", "VAL=\"World\""}
  
          lex := shell.NewLex(parser.DefaultEscapeToken)
          lex.RawQuotes = true
  
          val, err := lex.ProcessWord("Hello ${VAL}", envs)
          if err != nil {
                  t.Error(err)
          }   
  
          if val != "Hello World" {
                  t.Errorf("Expected Hello World but got %s", val)
          }   
  }
```

**Pull request merged**